File: sample/ps_java/boj/src/main/java/stack/BOJ1874.java
package stack;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class BOJ1874 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        Stack<Integer> st = new Stack<>();

        int[] arr = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(br.readLine());
        }
        StringBuilder sb = new StringBuilder();

        for (int i = 1, j = 0; i <= n; i++) {
            st.push(i);
            sb.append("+\n");
            if (st.peek() == arr[j]) {
                while (!st.isEmpty() && st.peek() == arr[j]) {
                    st.pop();
                    sb.append("-\n");
                    j++;
                }
            }
        }

        if (st.isEmpty())
            System.out.println(sb.toString());
        else
            System.out.println("NO");
    }
}

File: ps_java_kotlin/ps_kotlin/boj/src/main/kotlin/boj1874.kt
import java.util.*

fun main() {
    val br = System.`in`.bufferedReader()
    val sb = StringBuilder()

    val n = br.readLine().toInt()
    val st = Stack<Int>()
    var target = br.readLine().toInt()
    var readCount = 1

    for (i in 1..n) {
        st.push(i)
        sb.append("+\n")

        while (st.isNotEmpty() && st.peek() == target) {
            st.pop()
            if (readCount < n) {
                target = br.readLine().toInt()
                readCount++
            }
            sb.append("-\n")
        }

        if (st.isNotEmpty() && st.peek() > target) {
            break
        }
    }
    if (st.isEmpty()) print(sb.toString()) else println("NO")
}
